File: 4_kyu/Sum_of_Intervals.jl
function sumofintervals(a::Array{Tuple{T,T}})::T where {T<:Number}
    # (x,false) = start point at x, (x,true) = end point at x
    points = reduce( append!, map( p -> [(p[1],false),(p[2],true)], a ), init=Tuple{T,Bool}[] )
    sort!( points )
    sum = 0
    in_interval = 0
    start_interval = 0
    for p in points
        if p[2] # true = end point
            in_interval -= 1
            if in_interval == 0
                sum += p[1] - start_interval
            end
        else # false = start point
            in_interval += 1
            if in_interval == 1
                start_interval = p[1]
            end 
        end
    end        
    sum
end

File: 4_kyu/Sum_of_Intervals.py
def sum_of_intervals(intervals: list[list[int,int]]) -> int:
    """
    Calculate sum of interval lengths.
    Does not add overlapping intervals twice.
    """
    # Preparation step:
    #   Store start and end points in a sorted list.
    #   Store start point as (coordinate,+1) amd end point as (coordinate,-1)
    points = []
    for i in intervals:
        points.append((i[0],+1))
        points.append((i[1],-1))
    points.sort( key=lambda cf: cf[0] )
    # Calculate the sum of all interval lengths
    sum = 0 # sum of all closed intervals so far
    deep = 0 # number of current overlapping intervals
    start = 0 # start coordinate of current interval
    for p in points:
        if deep == 0 and p[1] == 1:
            # start of a new interval
            start = p[0]
        elif deep == 1 and p[1] == -1:
            # end of the current interval
            sum += p[0] - start
        deep += p[1]
    return sum
